projects/fal/src/fal/api.py

```diff
@@ -512,6 +548,47 @@ def run(
         return cast(ReturnT, return_value)
 
 
+    @_handle_grpc_error()
+    def run(
+        self,
+        func: Callable[..., ReturnT],
+        options: Options,
+        args: tuple[Any, ...],
+        kwargs: dict[str, Any],
+    ) -> ReturnT:
+        def result_handler(partial_result):
+            for log in partial_result.logs:
+                self._log_printer.print(log)
+
+        return self._run(func, options, args, kwargs, result_handler=result_handler)
+
+    def spawn(
+        self,
+        func: Callable[..., ReturnT],
+        options: Options,
+        args: tuple[Any, ...],
+        kwargs: dict[str, Any],
+    ) -> SpawnInfo:
+        ret = SpawnInfo()
+
+        def result_handler(partial_result):
+            for log in partial_result.logs:
+                if "Access your exposed service at" in log.message:
+                    ret.url = log.message.rsplit()[-1]
```

Comment:
yep, ugly, there will be a pr for controller to return URL (in both RegisterApplication and Run) from me later. Just want to get this going.